src/dbnode/encoding/proto/custom_unmarshal.go

```diff
@@ -332,14 +344,15 @@ func unmarshalSimpleField(fd *desc.FieldDescriptor, v uint64) (unmarshalValue, e
 }
 
 func (u *customUnmarshaler) resetAndUnmarshal(schema *desc.MessageDescriptor, buf []byte) error {
-	if schema != u.schema {
-		u.nonCustomValues = dynamic.NewMessage(schema)
-		u.nonCustomValues.Reset()
-	}
-
 	u.schema = schema
 	u.numNonCustom = 0
 	u.customValues = u.customValues[:0]
+
+	for i := range u.nonCustomValues {
+		u.nonCustomValues[i] = marshaledField{}
+	}
+	u.nonCustomValues = u.nonCustomValues[:0]
```

Comment:
You could make this an member method so in the other place you zero out you remember to zero the actual fields?  i.e. `u.resetNonCustomValues()`